org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -604,7 +604,7 @@ private boolean supportsWorkspaceFolderCapability() {
 				TextDocumentSyncKind syncKind = initializeFuture == null ? null
 						: serverCapabilities.getTextDocumentSync().map(Functions.identity(), TextDocumentSyncOptions::getChange);
 				final var listener = new DocumentContentSynchronizer(this, languageServer, theDocument, syncKind);
-				theDocument.addDocumentListener(listener);
+				theDocument.addPrenotifiedDocumentListener(listener);
```

Comment:
the javadoc of  `addPrenotifiedDocumentListener` and `removePrenotifiedDocumentListener` says 	

>  This method is not for public use.

Is that not something we need to respect?

Reply:
If not, DefaultDocumentAdapter could be used. It connect to Document via this API and expose TextChangeListener

Reply:
Or just tell adopters to not use IDocumentListener and migrate to ITextListener

Reply:
@mickaelistria , what do you think about this PR?

Reply:
I'm not much familiar with this API. If it ensures that `didChange` is called before the reconciliers are invoked, why not... I don't really see how it can hurt, particularly as it doesn't change the state of the IDocument.

Reply:
I also do not see how it can hurt. @zulus , could we follow on the comment "For now reconcilers have to wait couple ms to make sure they will be called after didChange notification" and change the reconcilers to not wait as part of this PR?

Reply:
@rubenporras as an example see #738 If IReconciler use standard path -> ITextViewer -> ITextInputListener -> IDocumentListener, his listener will be installed before LanguageServerWrapper, because both  ConnectDocumentToLanguageServerSetupParticipant and LanguageServerWrapper are async (awit for initialize, delayed excutor etc..)